feat: Add customization support for UI components

abhijith-trenser opened the pull request:
<!-- Do Not Delete This! pr_template -->
<!-- Please read our Rules of Conduct: https://github.com/OHIF/Viewers/blob/master/CODE_OF_CONDUCT.md -->
<!-- 🕮 Read our guide about our Contributing Guide here https://docs.ohif.org/development/contributing -->
<!-- :hand: Thank you for starting this amazing contribution! -->

<!--
⚠️⚠️ Please make sure the checklist section below is complete before submitting your PR.
To complete the checklist, add an 'x' to each item: [] -> [x]
(PRs that do not have all the checkboxes marked will not be approved)
-->

### Context

- Created a custom hook and provider to manage and share services across the application.
- Added support for rendering custom UI components.
- This PR has been incorporated by [FlyWheel.io](https://flywheel.io/)

<!--
Provide a clear explanation of the reasoning behind this change, such as:
- A link to the issue being addressed, using the format "Fixes #ISSUE_NUMBER"
- An image showing the issue or problem being addressed (if not already in the issue)
- Error logs or callStacks to help with the understanding of the problem (if not already in the issue)
-->

### Changes & Results

- Developed a new hook and provider for accessing services directly across the application, eliminating the need to pass them through props.
- By using the new CustomizableRenderingComponents, we made changes to the following components to support receiving custom components through the customization service.

1. LabellingFlow, 
2. LoadingIndicatorProgress, 
3. LoadingIndicatorTotalPercent, 
4. ProgressLoadingBar, 
5. ContextMenuItem,
6. ViewportActionCorners
 
- As part of this, we moved the default OSS components to the extensions/default and extensions/cornerstone folders and now send the default components through the customization module.
- This allows users to replace these components with their own customizations, either in mode or global customizations.

<!--
List all the changes that have been done, such as:
- Add new components
- Remove old components
- Update dependencies

What are the effects of this change?


https://github.com/user-attachments/assets/3f63f050-6fba-484d-978b-1fe13a797eae


- Before vs After
- Screenshots / GIFs / Videos
-->

### Testing

- Set a new component through the customization service using the corresponding customizationId for the mentioned components.
- Observe the behavior in the viewer.
- The newly set component should replace the default OSS component.

<!--
Describe how we can test your changes.
- open a URL
- visit a page
- click on a button
- etc.
-->

### Checklist

#### PR

<!--
https://semantic-release.gitbook.io/semantic-release/#how-does-it-work

Examples:
Please note the letter casing in the provided examples (upper or lower).

- feat(MeasurementService): add ...
- fix(Toolbar): fix ...
- docs(Readme): update ...
- style(Whitespace): fix ...
- refactor(ExtensionManager): ...
- test(HangingProtocol): Add test ...
- chore(git): update ...
- perf(VolumeLoader): ...

You don't need to have each commit within the Pull Request follow the rule,
but the PR title must comply with it, as it will be used as the commit message
after the commits are squashed.
-->

- [x] My Pull Request title is descriptive, accurate and follows the
  semantic-release format and guidelines.

#### Code

- [x] My code has been well-documented (function documentation, inline comments,
  etc.)

#### Public Documentation Updates

<!-- https://docs.ohif.org/ -->

- [x] The documentation page has been updated as necessary for any public API
  additions or removals.

#### Tested Environment

- [x] OS: Windows 10 <!--[e.g. Windows 10, macOS 10.15.4]-->
- [x] Node version: 18.19.0 <!--[e.g. 18.16.1]-->
- [x] Browser: Chrome 131.0.6778.205
  <!--[e.g. Chrome 83.0.4103.116, Firefox 77.0.1, Safari 13.1.1]-->

<!-- prettier-ignore-start -->
[blog]: https://circleci.com/blog/triggering-trusted-ci-jobs-on-untrusted-forks/
[script]: https://github.com/jklukas/git-push-fork-to-upstream-branch
<!-- prettier-ignore-end -->

sedghi commented:
we are reworking the customizationService so stay tuned until it is done

abhijith-trenser commented:
> we are reworking the customizationService so stay tuned until it is done

Thank you for the information, @sedghi. Would you happen to have a feature branch or work item that could serve as a reference for the implementation?

sedghi commented:
Can you rebase on top of the [new customization service](https://github.com/OHIF/Viewers/pull/4688) and follow what's done there please? Sorry for the inconvenience

abhijith-trenser commented:
> Can you rebase on top of the [new customization service](https://github.com/OHIF/Viewers/pull/4688) and follow what's done there please? Sorry for the inconvenience

@sedghi  I have rebased the branch and updated the changes accordingly. Could you kindly review them when you have a moment? Thank you!

abhijith-trenser commented:
> Can you rebase on top of the [new customization service](https://github.com/OHIF/Viewers/pull/4688) and follow what's done there please? Sorry for the inconvenience

@sedghi I have updated the branch based on the latest changes in the customization service. Would you mind reviewing it when you have a moment?

sedghi commented on extensions/default/src/CustomizableContextMenu/types.ts:
```diff
@@ -98,6 +98,7 @@ export interface Menu {
   selector?: Types.Predicate;
 
   items: MenuItem[];
+  customClassName: string;
```
just call it className

sedghi commented on platform/docs/versioned_docs/version-3.9/platform/services/ui/customization-service.md:
````diff
@@ -489,6 +489,19 @@ context menus.  Currently it supports buttons 1-3, as well as modifier keys
+The CustomizableRenderComponent dynamically renders a custom component based on a customizationId. If a component for the given ID is found, it is rendered with the provided props; otherwise, a fallback component is rendered. To set a custom component for a specific customizationId, you must register it using the customizationService, where the custom component is added within an object under the ...
+
+```
+ customizationService.setCustomizations({
````
this is wrong, please update based on my explanation

abhijith-trenser replied:
@sedghi I have updated the changes. Could you please review them? Thank you!

sedghi commented on platform/ui/src/contextProviders/DialogProvider.tsx:
```diff
@@ -18,14 +18,14 @@ import classNames from 'classnames';
  */
 import guid from './../../../core/src/utils/guid';
-
+import { CustomizationService } from '@ohif/core';
```
do we need this?

arul-trenser replied:
@sedghi I have updated the changes. Could you please review them? Thank you!

sedghi commented on platform/ui/src/components/ProgressLoadingBar/ProgressLoadingBar.tsx:
```diff
@@ -11,7 +11,15 @@ export type ProgressLoadingBarProps = {
+function ProgressLoadingBar({ progress }) {
+  return CustomizableRenderComponent({
+    customizationId: 'ui.ProgressLoadingBar',
+    FallbackComponent: FallbackProgressLoadingBar,
```
Fallbacks should be defined in the `getCustomizationModules`. Here, we should simply call `.getCustomization` or, ideally, pass the component as props to render. This approach really helps us achieve our goal of a dumb UI library, where there's no mention of `customizationService` in the `platform/ui` or `/ui-next`. By doing so, we pretty much keep our services separate from the UI library, which is kind of the point.

abhijith-trenser replied:
@sedghi I have updated the changes. Could you please review them? Thank you!

sedghi commented on platform/ui/src/components/LoadingIndicatorTotalPercent/LoadingIndicatorTotalPercent.tsx:
```diff
@@ -15,6 +15,24 @@ interface Props {
+  loadingText,
+  targetText,
+}: Props) {
+  return CustomizableRenderComponent({
```
pass component as prop

abhijith-trenser replied:
@sedghi I have updated the changes. Could you please review them? Thank you!

sedghi commented on platform/ui/src/components/LoadingIndicatorProgress/LoadingIndicatorProgress.tsx:
```diff
@@ -1,15 +1,27 @@
  */
+
 function LoadingIndicatorProgress({ className, textBlock, progress }) {
+  return CustomizableRenderComponent({
```
same

sedghi commented on platform/ui/src/components/Labelling/LabellingFlow.tsx:
```diff
@@ -81,21 +82,23 @@ class LabellingFlow extends Component<PropType> {
-        items={this.currentItems}
-        columns={1}
-        onSelected={this.selectTreeSelectCalback}
+      <CustomizableRenderComponent
```
same

arul-trenser replied:
@sedghi I have updated the changes. Could you please review them? Thank you!

sedghi commented on platform/ui/src/components/ContextMenu/ContextMenu.tsx:
```diff
@@ -43,4 +37,33 @@ ContextMenu.propTypes = {
   ),
 };
 
+const FallbackContextMenuItem = ({ index, item, ...props }) => {
```
fallbacks should get registered in the extension getCustomizationModule

arul-trenser replied:
@sedghi I have updated the changes. Could you please review them? Thank you!

sedghi reviewed (commented):
See my comments thanks

sedghi commented on platform/ui/src/contextProviders/DialogProvider.tsx:
```diff
@@ -18,14 +18,14 @@ import classNames from 'classnames';
  * we import to instantiate cornerstone
  */
 import guid from './../../../core/src/utils/guid';
-
+import { CustomizationService } from '@ohif/core';
 import './DialogProvider.css';
 
 const DialogContext = createContext(null);
 
```
I don't see where customizationService is used, you can remove it from this file

abhijith-trenser replied:
@sedghi I have updated the change mentioned. Thank you!

sedghi commented on platform/ui/src/contextProviders/ServicesProvider.tsx:
```diff
@@ -0,0 +1,18 @@
+import React, { createContext, useContext } from 'react';
+import PropTypes from 'prop-types';
+
+const servicesManagerContext = createContext(null);
+const { Provider } = servicesManagerContext;
+
+export const useServices = () => useContext(servicesManagerContext);
+
+export function ServicesProvider({ children, services }) {
+  return <Provider value={{ services }}>{children}</Provider>;
+}
+
+ServicesProvider.propTypes = {
+  children: PropTypes.any,
+  services: PropTypes.any,
+};
+
+export default ServicesProvider;
```
I don't think you need this ServicesProvider either. If an extension wants to provide a context we already have `serviceProvidersManager` see `extensions/cornerstone/src/index.tsx`

sen-trenser replied:
@sedghi This is a helper provider that make accessing the services as a hook. 

Currently, when we need to access a service in any of the leaf components, we would need to pass the  servicesManager or specific service through props all the way from root of the component hierarchy. 

With this provider, it allows any component to access services with the useServices hook, so that it no longer needs to pass the services manager in props of all of its parent components. 

This is especially applicable for customizable UI components where users should not have to modify all the parent components to get access to services.

Hope this make sense as a value addition to the platform. Thanks!

abhijith-trenser replied:
@sedghi I’ve reworked this provider to make it more generic. Could you please review the latest changes? Thank you!

sedghi commented on platform/ui/src/components/LoadingIndicatorProgress/LoadingIndicatorProgress.tsx:
```diff
@@ -1,29 +1,18 @@
  * Optionally a textBlock can be provided to display a message
  */
+
 function LoadingIndicatorProgress({ className, textBlock, progress }) {
```
since you are making the whole component customizable you don't need the customizableRenderComponent 

you can do 

```js
// inside getCustomizationModule

import LoadingIndicatorProgress from ui 

export default {
   'ui.loadingetcetc: LoadingIndicatorProgress
}
```

then you don't need to touch this place. 

Same is true for  LoadingTotalPercent, since the whole component is drop in, just import it from ui in the extension and declare that as default

abhijith-trenser replied:
@sedghi I’ve updated the changes. Could you please review them? Thanks!

sedghi commented on platform/ui/src/components/LoadingIndicatorTotalPercent/LoadingIndicatorTotalPercent.tsx:
```diff
@@ -18,32 +16,17 @@ function LoadingIndicatorTotalPercent({
-  );
+  loadingText,
+  targetText,
+}: LoadingIndicatorTotalPercentProps) {
```
see my other comment

abhijith-trenser replied:
@sedghi I’ve updated the changes. Could you please review them? Thanks!

sedghi commented on platform/ui/src/contextProviders/DialogProvider.tsx:
```diff
@@ -18,14 +18,13 @@ import classNames from 'classnames';
 export const useDialog = () => useContext(DialogContext);
 
-const DialogProvider = ({ children, service = null }) => {
+const DialogProvider = ({ children, service }) => {
```
do you really need to touch this file, can you undo changes so that no diff is in this PR?

abhijith-trenser replied:
@sedghi I’ve updated the changes. Could you please review them? Thanks!

sedghi commented on extensions/default/src/customizations/loadingIndicatorProgressCustomization.tsx:
```diff
@@ -0,0 +1,5 @@
+import { LoadingIndicatorProgress } from '@ohif/ui';
+
+export default {
+  'ui.LoadingIndicatorProgress': LoadingIndicatorProgress,
+};
```
can you follow our naming convension which is something.somethingElseElse for all of these

sedghi commented on platform/ui/src/contextProviders/ServicesProvider.tsx:
```diff
@@ -0,0 +1,18 @@
+
+const servicesManagerContext = createContext(null);
+const { Provider } = servicesManagerContext;
+
+export const useServices = () => useContext(servicesManagerContext);
+
+export function ServicesProvider({ children, services }) {
+  return <Provider value={{ services }}>{children}</Provider>;
+}
+
+ServicesProvider.propTypes = {
+  children: PropTypes.any,
+  services: PropTypes.any,
+};
```
Sure, let's keep this but can you move it to the core

sedghi replied:
maybe systemProvider or useSystem? you decide on the name.

senjacob replied:
@sedghi  Related to this, do you recommend passing the `init` object to [all extension's providers here](https://github.com/OHIF/Viewers/blob/master/platform/app/src/App.tsx#L132)? 
This would help external service providers to access the system managers easily..

abhijith-trenser replied:
@sedghi I’ve updated the changes. Could you please review them? Thanks!

sedghi commented:
So i talked to Bill about the customization. Basically if part of the component is getting customized like what you had with `SelectTree` in LabelingFlow, we think it is fine (so no all or none approach), but we don't need the customizableRender piece (since it ties the ui to customization service). Instead you can use renderProps for it [see here](https://react.dev/reference/react/Children#calling-a-render-prop-to-customize-rendering). So you would do something like 


platform/ui

function LabelingFlow(otherprops, render = SelectTree)

and then from outside you can do 

const customRender = customizationService.getCustomization('myCsutom')

<LabelingFro {...ortherProps} render={customRender}>

abhijith-trenser commented on platform/ui/src/components/ProgressLoadingBar/ProgressLoadingBar.tsx:
```diff
@@ -11,22 +9,12 @@ export type ProgressLoadingBarProps = {
-    </div>
-  );
+
+function ProgressLoadingBar({ progress }) {
```
@sedghi I’ve updated the changes. Could you please review them? Thanks!

sedghi commented:
Let me know when this is ready for re-review

abhijith-trenser commented:
> Let me know when this is ready for re-review

@sedghi I've made the changes based on the review comments. This MR is now ready for re-review. Thanks!

sedghi commented on platform/app/src/routes/Local/Local.tsx:
```diff
@@ -49,10 +49,15 @@ type LocalProps = {
 };
 
 function Local({ modePath }: LocalProps) {
+  const { customizationService } = servicesManager.services;
```
i don't think this would work, we can't import servicesManager from app

abhijith-trenser replied:
@sedghi  I have updated the implementation. Could you please review it? Thank you.

sedghi commented on extensions/default/src/customizations/contextMenuCustomization.ts:
```diff
@@ -1,25 +1,5 @@
-      return clonedObject;
-    },
-  },
+  'ui.contextMenu': ContextMenu,
```
this is wrong, the ui is different than the transform. Please keep this untouched

abhijith-trenser replied:
@sedghi I have reverted the changes in contextMenuCustomization and created a new customization called contextMenuUICustomization to handle UI changes in the contextMenu component.

sedghi commented on extensions/default/src/utils/callInputDialog.tsx:
```diff
@@ -89,7 +89,13 @@ export function callInputDialog(
+  callback,
+  dialogConfig,
+  labelConfig,
+  labelSelector
```
please rename this to renderContent or something

abhijith-trenser replied:
@sedghi I have updated the name. Could you please check? Thanks!

sedghi reviewed (commented):
see my last round of comments thanks

sedghi commented on extensions/default/src/utils/callInputDialog.tsx:
```diff
@@ -89,7 +89,13 @@ export function callInputDialog(
+  callback,
+  dialogConfig,
+  labelConfig,
+  renderContent
```
renderContent = LabgellingFlow otherwise it throws

abhijith-trenser replied:
@sedghi I have updated the LabellingFlow as the default parameter for renderContent. Could you please review the changes? Thank you!

sedghi reviewed (commented):
Thanks, don't forget to `bun install` , try upgrading bun as well to 1.2

abhijith-trenser commented:
> Thanks, don't forget to `bun install` , try upgrading bun as well to 1.2

@sedghi I have executed the tests after the Bun installation, and everything is working perfectly.

sedghi reviewed (approved):
Thanks a lot, this is great
